smack-experimental/src/main/java/org/jivesoftware/smackx/reactions/ReactionsManager.java

```diff
@@ -0,0 +1,364 @@
+ * It also allows propagating these restrictions to other clients via XMPP service discovery.
+ *
+ * This class is based on the XEP-0444 extension protocol for reactions.
+ *
+ * @author Ismael Nunes Campos
+ *
+ * @see <a href="https://xmpp.org/extensions/xep-0444.html">XEP-0444 Message Reactions</a>
+ * @see ReactionsElement
+ * @see Reaction
+ */
+public final class ReactionsManager extends Manager {
+
+    private static final Map<XMPPConnection, ReactionsManager> INSTANCES = new WeakHashMap<>();
+
+    static {
+        XMPPConnectionRegistry.addConnectionCreationListener(new ConnectionCreationListener() {
+            @Override
+            public void connectionCreated(XMPPConnection connection) {
+                getInstanceFor(connection);
+            }
+        });
+    }
+
+    private static final String REACTIONS_RESTRICTIONS_NAMESPACE = "urn:xmpp:reactions:0:restrictions";
+    private final Set<ReactionsListener> listeners = new CopyOnWriteArraySet<>();
+    private final AsyncButOrdered<BareJid> asyncButOrdered = new AsyncButOrdered<>();
+    private final StanzaFilter reactionsElementFilter = new AndFilter(StanzaTypeFilter.MESSAGE, ReactionsFilter.INSTANCE);
+
+    /**
+     * Constructs an instance of the reactions manager and add ReactionsElement to disco features.
+     *
+     * @param connection The XMPP connection used by the manager.
+     */
+    public ReactionsManager(XMPPConnection connection) {
+        super(connection);
+        connection.addAsyncStanzaListener(this::reactionsElementListener, reactionsElementFilter);
+        ServiceDiscoveryManager sdm = ServiceDiscoveryManager.getInstanceFor(connection);
+        sdm.addFeature(ReactionsElement.NAMESPACE);
+    }
+
+    /**
+     * Listener method for reactions elements in XMPP messages. This method is invoked when a new
+     * stanza (message) is received and attempts to extract a {@link ReactionsElement} from the message.
+     * If the element is found, it notifies the registered reaction listeners.
+     *
+     * @param packet The received XMPP stanza (message).
+     */
+    public void reactionsElementListener(Stanza packet) {
+        Message message = (Message) packet;
+        ReactionsElement reactionsElement = ReactionsElement.fromMessage(message);
+        if (reactionsElement != null) {
+            notifyReactionListeners(message, reactionsElement);
+        }
+    }
+
+    /**
+     * Notifies all registered reaction listeners that a new reaction has been received. This method
+     * performs the notification in an ordered, asynchronous manner to ensure listeners are notified in
+     * the order that they were added.
+     *
+     * @param message The XMPP message that contains the reactions.
+     * @param reactionsElement The {@link ReactionsElement} containing the reactions.
+     */
+    public void notifyReactionListeners(Message message, ReactionsElement reactionsElement) {
```

Comment:
Why is this method `public`? Why does it exist at all if there is only exactly one call site?